src/expiring_lru.rs

```diff
@@ -0,0 +1,264 @@
+where
+    K: Eq + Hash,
+{
+    let mut interval = interval(reap_interval);
+
+    loop {
+        interval.tick().await;
+
+        let right_now = Instant::now();
+        map.inner
+            .retain(|_, entry| entry.last_used + expiration > right_now);
+
+        debug!("Done reaping timed out HTTP ratelimiters");
+    }
+}
+
+pub struct ExpiringLru<K, V> {
+    inner: Arc<DashMap<K, Entry<V>>>,
+    max_size: Option<usize>,
+}
+
+impl<K, V> Clone for ExpiringLru<K, V> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            max_size: self.max_size,
+        }
+    }
+}
+
+impl<K, V> ExpiringLru<K, V>
+where
+    K: Eq + Hash + Clone + Send + Sync + 'static,
+    V: Send + Sync + 'static,
+{
+    fn new(reap_interval: Duration, expiration: Duration, max_size: Option<usize>) -> Self {
+        let inner = Arc::new(DashMap::new());
+
+        let this = Self { inner, max_size };
+
+        tokio::spawn(expire_entries(this.clone(), reap_interval, expiration));
+
+        this
+    }
+
+    pub fn insert(&self, key: K, value: V) {
+        match self.max_size {
+            Some(max_size) if max_size == 0 => return,
+            Some(max_size) if self.len() >= max_size => {
+                if let Some(lru) = self.get_lru() {
+                    let key = lru.key().clone();
+                    // We can't hold any references when removing something from the map
+                    drop(lru);
+                    self.remove(&key);
+
+                    debug!("Removed least recently used entry from expiring LRU");
+                }
+            }
+            _ => {}
+        }
+
+        let last_used = Instant::now();
+        let entry = Entry {
+            inner: value,
+            last_used,
+        };
+        self.inner.insert(key, entry);
+    }
+
+    pub fn get<Q>(&self, key: &Q) -> Option<EntryRef<'_, K, V>>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        self.inner.get_mut(key).map(|mut entry| {
+            entry.last_used = Instant::now();
+
+            EntryRef(EntryRefInner::RefMut(entry))
+        })
+    }
+
+    pub fn get_lru(&self) -> Option<EntryRef<'_, K, V>> {
+        self.inner
+            .iter()
+            .next()
+            .map(|first_entry| {
+                self.inner.iter().fold(first_entry, |old, next| {
+                    if old.last_used > next.last_used {
+                        next
+                    } else {
+                        old
+                    }
+                })
+            })
+            .map(|entry| EntryRef(EntryRefInner::RefMulti(entry)))
+    }
+
+    pub fn remove(&self, key: &K) -> Option<(K, Entry<V>)> {
+        self.inner.remove(key)
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn full(&self) -> bool {
+        match self.max_size {
+            Some(max_size) => self.len() >= max_size,
```

Comment:
I'd rather have it panic should len >= max_size to catch bugs instead of silently continue "working"